### ISSUE TITLE
cgo,native: consider early checkMatrix and checkVector

### ISSUE THREAD

**Issue by kortschak** [gonum/lapack#228]
In some cgo and native implementations, checkMatrix and checkVector occur after work length query. I have been writing them very early under the Principle of Repair, but we should be behaving consistently.

This issue is to discuss which way we should change the existing code; either to allow work length queries without checks or require all parameters be correct under all conditions.

**Comment by vladimir-ch** [comment]
My intention with skipping the checks when `lwork == -1` was to make possible to call the function with nil slices, so that it is explicit in the caller that the optimal workspace size is determined only based on the other parameters, not the slice length. Since the caller may still provide that slices, currently the code is not consistent in this. Requiring all parameters be correct under all conditions is simpler and will inevitably lead to consistency. This seems like a better option.

**Comment by vladimir-ch** [comment]
A case against requiring all parameters be correct under all conditions is `Dgesvd`. There we need to allocate an mxn slice just for the work query.

**Comment by kortschak** [comment]
True. Similarly in `Dggsvp3` I check `tau` late. I think this is something that should not be rigid, but should happen as early as possible.

**Comment by vladimir-ch** [comment]
I think we should get rid of that allocation in Dgesvd but it seems that all the functions used in the work query would have to be modified. Or do you see another way?
Speaking of allocations, there is another `make` in `Dormlq`, it would be nice to remove that too.

**Comment by kortschak** [comment]
A quick look at `Dgeqrf`. If they are all like that, I think we can weaken the requirements to how they are in `Dggsvp3`. In the case of `Dgeqrf`, `tau` is not checked until after `lwork != -1` and the length of work need only be at least 1 until then.

I can go through and do those.

**Comment by kortschak** [comment]
I don't see a way to avoid the allocation in `Dormlq`.
